server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalance.java

```diff
@@ -30,4 +31,70 @@ public static boolean hasChanges(DesiredBalance a, DesiredBalance b) {
         return Objects.equals(a.assignments, b.assignments) == false;
     }
 
+    public int unassigned() {
+        int unassigned = 0;
+        for (ShardAssignment assignment : assignments.values()) {
+            unassigned += assignment.unassigned();
+        }
+        return unassigned;
+    }
+
+    public int total() {
+        int total = 0;
+        for (ShardAssignment assignment : assignments.values()) {
+            total += assignment.total();
+        }
+        return total;
+    }
+
+    public static int shardMovements(DesiredBalance old, DesiredBalance updated) {
+        var intersection = Sets.intersection(old.assignments().keySet(), updated.assignments().keySet());
+        int movements = 0;
+        for (ShardId shardId : intersection) {
+            var oldAssignment = old.getAssignment(shardId);
+            var updatedAssignment = updated.getAssignment(shardId);
+            if (Objects.equals(oldAssignment, updatedAssignment) == false) {
+                movements += shardMovements(oldAssignment, updatedAssignment);
+            }
+        }
+        return movements;
+    }
+
+    public static int shardMovements(ShardAssignment old, ShardAssignment updated) {
```

Comment:
Could we make this `private`? It's only exposed for tests AFAICT and we could test it through the one that compares at the `DesiredBalance` level instead.